### PR TITLE
HPCC-18805 Ensure the LHS of a FETCH is in the serialized format

### PR DESCRIPTION
**Pull request by ghalliday**
Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [ ] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:

Ran full compiler regression suite and checked output.  Checked that the problem fixed Richard's initial issue.  (His test will be released in a separate PR.)
Note ecl/regress/fetch11.ecl changes as a result of this - the previous code was incorrect.

### CODE REVIEW

**Comment by ghalliday** [comment]
@shamser please review.

File: ecl/hql/hqlopt.cpp [hunk 1]
```diff
@@ -3786,6 +3786,18 @@ IHqlExpression * CTreeOptimizer::doCreateTransformed(IHqlExpression * transforme
...
             {
+            //None of the follow change the format of the left hand side, and since the fetch is on the underlying
+            //dataset they can all be removed.  If they remain the disk file can be treated as if it is in memory format.
+            case no_preload:
+            case no_nofold:
+            case no_nocombine:
+            case no_dedup:
+            case no_group:
+            case no_compound_diskread:
+            case no_filter:
```
Comment (shamser) [inline ecl/hql/hqlopt.cpp]: @ghalliday Is no_dedup and no_filter valid operation on the dataset used in a fetch.  If not shouldn't it throw an exception?

File: ecl/hqlcpp/hqlsource.cpp [hunk 2]
```diff
@@ -7253,6 +7253,12 @@ ABoundActivity * HqlCppTranslator::doBuildActivityFetch(BuildCtx & ctx, IHqlExpr
         OwnedHqlExpr simple = replaceFetchInput(expr, null, no_right);
         OwnedHqlExpr transformed = replaceExpression(simple, tableExpr, projected);
         OwnedHqlExpr optSimple = optimizeHqlExpression(queryErrorProcessor(), transformed, optFlags);
+        IHqlExpression * fetch = queryFetch(optSimple);
```
Comment (shamser) [inline ecl/hqlcpp/hqlsource.cpp]: @ghalliday minor: may be better not to have this 'fetch' variable shadow the 'fetch' variable in parent scope.

**Review by shamser — COMMENTED** [review]
@ghalliday One question and one comment.

File: ecl/hql/hqlopt.cpp [hunk 3]
```diff
@@ -3786,6 +3786,18 @@ IHqlExpression * CTreeOptimizer::doCreateTransformed(IHqlExpression * transforme
             node_operator childOp = child->getOperator();
             switch(childOp)
             {
+            //None of the follow change the format of the left hand side, and since the fetch is on the underlying
```
Comment (shamser) [inline ecl/hql/hqlopt.cpp]: @ghalliday "None of the _following_ change..."

File: ecl/hql/hqlopt.cpp (hunk 1 above)
Comment (ghalliday) [reply ecl/hql/hqlopt.cpp]: Apparently it is - I have test cases which check it is supported.  It probably shouldn't be, but I suspect it was in response to a bug report. (fetch11)

File: ecl/hqlcpp/hqlsource.cpp (hunk 2 above)
Comment (ghalliday) [reply ecl/hqlcpp/hqlsource.cpp]: agreed, I will fix.

**Comment by ghalliday** [comment]
@richardkchapman changes suggested by shamser included, and commit squashed

**Pull request merged**